shared/middleware/helpers/bucket_helper.rb

```diff
@@ -108,19 +108,20 @@ def copy_files(src_channel, dest_channel, options={})
     dest_owner_id, dest_channel_id = storage_decrypt_channel_id(dest_channel)
 
     src_prefix = s3_path src_owner_id, src_channel_id
-    @s3.list_objects(bucket: @bucket, prefix: src_prefix).contents.map do |fileinfo|
+    result = @s3.list_objects(bucket: @bucket, prefix: src_prefix).contents.map do |fileinfo|
       filename = %r{#{src_prefix}(.+)$}.match(fileinfo.key)[1]
-      if !options[:filenames] || options[:filenames].include?(filename)
+      if (!options[:filenames] && (!options[:exclude_filenames] || !options[:exclude_filenames].include?(filename))) || options[:filenames].try(:include?, filename)
```

Comment:
Good point. It took me a while to make sure I had it right as well. I'll make a note to extract a helper after I merge these changes.